### ISSUE TITLE
StackedEmbeddings modifies original embeddings' name which causes unwanted behaviour and potential crashes

### ISSUE THREAD

**Issue by tadejmagajna** [flairNLP/flair#2600]
**Description**
When `StackedEmbeddings` are initialised, they modify the input embeddings by renaming their `name` property as part of this code:

https://github.com/flairNLP/flair/blob/579c7b70dfcc1d184bd47069f722d7a9ae8b78d7/flair/embeddings/token.py#L49-L50

The above code modifies the original embeddings' name. But several parts of Flair rely on embedding names being consistent.

For example the [`_everything_embedded()`](https://github.com/flairNLP/flair/blob/b774774752c8338aab3d620f7e5062f66ec7a69d/flair/embeddings/base.py#L67) method relies on embedding names when checking whether all embeddings were applied. When an embedding's name changes, its methods like `_everything_embedded()` become broken and this potentially results in the same embedding being added twice to the same token.

The adverse conesquences of this bug and the consequences follow this pattern:

1. The original embeddding's (let's call it `Char`) gets renamed to `0-Char` the first time we call StackedEmbeddings. If we pass the same embedding object to another StackedEmbeddings class it'll be renamed to `0-0-Char` and so on...
2. `_everything_embedded()` method breaks and **even if all tokens in a `Sentence` are embedded, the method will think they aren't** because the embedding now has a new name and no tokens were embedded with an embedding that has this new name. This results in the token being embedded again the next time we call `.embed()` on the same token resulting in duplicated embedding dimensionality.
3. The tokens are now embedded with double embedding dimensionality which causes a number of downstream issues and crashes

**To Reproduce**
```python
from flair.embeddings import StackedEmbeddings, CharacterEmbeddings
from flair.data import Sentence

# setup
s = Sentence('a')
e = CharacterEmbeddings(hidden_size_char=2)
e.embed(s)
print("name of the embedding is", e.name)  # name is "Char"
print("embedding len is:", len(s[0].embedding))  # embedding is 4-dimentional
e.embed(s)
print("embedding len is:", len(s[0].embedding))  # embedding is 4-dimentional regardless of how many times we call embed()

# trigger the bug
stack = StackedEmbeddings([e])
e.embed(s)

# unwanted results
print("name of the embedding is now:", e.name)  # name is "0-Char"
print("embedding len is now:", len(s[0].embedding))  # embedding is now 8-dimentional
```

**Expected behaviour**
The way `StackedEmbeddings` modifies the original embeddings can cause a number of issues if we plan to use the original embedding object later on in any way.

It appears to me that this modified embedding name is only used so that it can be passed to the Torch `add_module()` method. Instead of renaming the original embeddings' name itself, we could simply pass the modified embedding name to the `add_module()` function directly like we do in: https://github.com/flairNLP/flair/blob/7d5746fa072ccf186f27c720c0dbc93f62622dae/flair/embeddings/document.py#L443

Happy to open a PR with the fix!

**Comment by alanakbik** [comment]
@tadejmagajna makes sense! Probably the OrderedDict solution is best since this does not require us to increase the Python version requirements. (Though probably going to 3.7 has to happen at some point)
